gateway/src/apicast/policy_loader.lua

```diff
@@ -0,0 +1,198 @@
+--- Policy loader
+-- This module loads a policy defined by its name and version.
+-- It uses sandboxed require to isolate dependencies and not mutate global state.
+
+local format = string.format
+local error = error
+local type = type
+local loadfile = loadfile
+local getenv = os.getenv
+local insert = table.insert
+local setmetatable = setmetatable
+local concat = table.concat
+
+local _M = { }
+
+local searchpath = package.searchpath
+local root_loaded = package.loaded
+
+local root_require = require
+local pcall = pcall
+
+local preload = package.preload
+
+preload['apicast.policy_loader'] = function() return _M end
+
+local prequire = function(...) return pcall(root_require, ...) end
+
+--- create a require function not using the global namespace
+-- loading code from policy namespace should have no effect on the global namespace
+-- but poliocy can load shared libraries that would be cached globally
+local function gen_require(package)
+
+  local function not_found(modname, err)
+    return error(format("module '%s' not found:%s", modname, err), 0)
+  end
+
+  --- helper function to safely use the native require function
+  local function fallback(modname, err)
+    local mod
+
+    mod = package.loaded[modname]
+
+    if not mod then
+      local ok
+      ngx.log(ngx.DEBUG, 'native require for: ', modname)
+      ok, mod = prequire(modname)
+
+      if not ok then return not_found(modname, err) end
+    end
+
+    return mod
+  end
+
+  --- helper function to find and return correct loader for a module
+  local function find_loader(modname)
+    local loader, file, err, ret
+
+    for i=1, #package.searchers do
+      ret, err = package.searchers[i](modname)
+
+      if type(ret) == 'function' then
```

Comment:
Would be good to paste parts of the Lua documentation describing what loaders can return.